From 7d84c165528548b50d6029d351be7b8899eaffba Mon Sep 17 00:00:00 2001
From: Venkatasaiprudhvi Sannidhi <vsannidhi@maxlinear.com>
Date: Mon, 9 Sep 2024 19:44:08 +0530
Subject: [PATCH] WLANRTSYS-61674 Duplication of RNR IE info when MLO is
 configured Issue: In case of MBSS, when MLO is established with a VAP which
 is part of MBSS, there are 2 Neighbor AP Information entries for the same
 VAP. The first Neighbor AP Information consists of TBTT information with MLD
 Parameters, and the other Neighbor AP Information contains TBTT information
 without MLD Parameters.

Fix: While preparing the RNR IE (in mxl_create_rnr_vap), check if the
colocated 6G VAP is part of ML and dont prepare the Neighbor info for it
as it is already prepared in mxl_create_rnr_mld.
---
 src/ap/mxl_beacon.c | 17 ++++++++++-------
 1 file changed, 10 insertions(+), 7 deletions(-)

diff --git a/src/ap/mxl_beacon.c b/src/ap/mxl_beacon.c
index 9948c5078..1465ff924 100644
--- a/src/ap/mxl_beacon.c
+++ b/src/ap/mxl_beacon.c
@@ -311,6 +311,7 @@ static void mxl_create_rnr_tbtt_hdr(u8 **pos, int radio, struct hostapd_data *ha
 static u8 * mxl_create_rnr_vap(u8 **pos, int radio, struct hostapd_data *hapd, bool is_same_ssid, u8 *len, struct ieee80211_neighbor_ap_info *rnr)
 {
 	u8 empty_bssid[ETH_ALEN] = {'\0'};
+	u8 temp_len = *len;
 	size_t j, count = 0;
 	struct ieee80211_neighbor_vap_info *rnr_vap;
 	u8 *temp_pos = *pos;
@@ -340,15 +341,11 @@ static u8 * mxl_create_rnr_vap(u8 **pos, int radio, struct hostapd_data *hapd, b
 			}
 			if (matchfound)
 				continue;
-#if 0
-		#ToDo: Need to optimize the rnr ie wrt MLD
-		} else if (hapd->conf->mlo_enable && hapd->sibling_hapd) {
-			if (!os_memcmp(hapd->iconf->coloc_6g_ap_info[radio].bss_info[j].bssid,
-						hapd->sibling_hapd->own_addr, ETH_ALEN)) {
+		} else if (hapd->conf->mxl_bss_conf.mlo_enable && hapd->mxl_data.sibling_hapd) {
+			if (!os_memcmp(hapd->iconf->mxl_conf.coloc_6g_ap_info[radio].bss_info[j].bssid,
+						hapd->mxl_data.sibling_hapd->own_addr, ETH_ALEN)) {
 				continue;
 			}
-#endif
-
 		}
 #endif /* CONFIG_IEEE80211BE_MXL_MLO */
 
@@ -403,6 +400,12 @@ static u8 * mxl_create_rnr_vap(u8 **pos, int radio, struct hostapd_data *hapd, b
 		count++;
 	}
 	rnr->tbtt_info_hdr |= ((count - 1) << RNR_TBTT_INFO_COUNT_OFFSET);
+	/* don't add empty TBTT Header */
+	if (*len == temp_len) {
+		*len -= sizeof(*rnr);
+		*pos -= sizeof(*rnr);
+		os_memset(*pos, 0, sizeof(*rnr));
+	}
 	return len;
 }
 
-- 
2.43.0

